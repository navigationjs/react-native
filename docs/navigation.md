---
id: navigation
title: Navigation
---

## Example

Example

### `name`

Name
