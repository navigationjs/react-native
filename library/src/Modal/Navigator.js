export default class Navigator {
  constructor(name) {
    this.name = name;
    this.scenes = {};
    this.history = [];
  }

  addScenes = (...scenes) => {
    scenes.forEach(it => (this.scenes[it.name] = it));
  };

  current = () => this.history[this.history.length - 1];

  go = async (name, duration) => {
    const scene = this.scenes[name];
    if (!scene) return Promise.reject();
    const alreadyInHistory = this.history.includes(name);
    if (alreadyInHistory) return Promise.resolve();
    await scene.show(duration);
    this.history.push(name);
  };

  back = async duration => {
    if (this.history.length === 0) return Promise.resolve();
    const name = this.current();
    const scene = this.scenes[name];
    if (!scene) return Promise.reject();
    await scene.hide(duration);
    this.history.pop();
  };

  reset = async () => {
    await Promise.all(
      Object.keys(this.scenes).map(key => this.scenes[key].hide(0))
    );
    this.history = [];
  };
}
