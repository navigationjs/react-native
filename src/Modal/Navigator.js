import { History } from '@navigationjs/core';

export default class Navigator {
  constructor(name) {
    this.name = name;
    this.scenes = {};
    this.history = new History(this.name);
  }

  addScenes = (...scenes) => {
    scenes.forEach(it => (this.scenes[it.name] = it));
  };

  removeScenes = (...scenes) => {
    scenes.forEach(it => delete this.scenes[it]);
  };

  current = () => this.history.current();

  go = async (name, duration) => {
    const scene = this.scenes[name];
    if (!scene) return Promise.reject();
    const alreadyInHistory = this.history.current() === name;
    if (alreadyInHistory) return Promise.resolve();
    await scene.show(duration);
    this.history.push(name);
  };

  back = async duration => {
    if (this.history.isEmpty()) return Promise.resolve();
    const name = this.current();
    const scene = this.scenes[name];
    if (!scene) return Promise.reject();
    await scene.hide(duration);
    this.history.pop();
  };

  reset = async () => {
    await Promise.all(
      Object.keys(this.scenes).map(key => this.scenes[key].hide(0))
    );
    this.history = new History(this.name);
  };
}
