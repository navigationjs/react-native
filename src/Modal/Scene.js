import Value from '../Value';

export default class Scene {
  constructor(name) {
    this.name = name;
    this.active = new Value('active');
  }

  show = duration => this.active.to(1, duration);
  hide = duration => this.active.to(0, duration);
}
