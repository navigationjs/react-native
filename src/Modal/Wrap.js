import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';

const { height } = Dimensions.get('window');

export default class Wrap extends Component {
  state = { loading: this.props.scene.active.loading };

  componentDidMount() {
    const {
      scene: { active },
    } = this.props;
    active.value.addListener(() => this.setState({ loading: active.loading }));
  }

  render() {
    const { scene, children, overlay, style, ...props } = this.props;
    const { loading } = this.state;

    return (
      <>
        {overlay && (
          <Animated.View
            style={{
              position: 'absolute',
              height: '100%',
              width: '100%',
              overflow: 'hidden',
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 0, 1],
                    outputRange: [height, 0, 0],
                  }),
                },
              ],
              backgroundColor: scene.active.value.interpolate({
                inputRange: [0, 1],
                outputRange: ['rgba(0,0,0,0)', 'rgba(0,0,0,0.4)'],
              }),
            }}
          />
        )}
        <Animated.View
          {...props}
          style={[
            {
              position: 'absolute',
              width: '100%',
              height: '100%',
              backgroundColor: 'white',
              overflow: 'hidden',
            },
            {
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 1],
                    outputRange: [height, 0],
                  }),
                },
              ],
            },
            style,
          ]}
        >
          {typeof children === 'function' ? children({ loading }) : children}
        </Animated.View>
      </>
    );
  }
}
