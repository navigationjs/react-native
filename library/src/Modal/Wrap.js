import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';
import { toId } from '../helpers';
import navigation from '../Navigation';

const { height } = Dimensions.get('window');

export default class Wrap extends Component {
  constructor(props) {
    super(props);

    const { navigator: navigatorName, scene: sceneName } = props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    this.state = { loading: scene.active.loading };
  }

  componentDidMount() {
    const { navigator: navigatorName, scene: sceneName } = this.props;
    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];
    scene.active.value.addListener(() =>
      this.setState({ loading: scene.active.loading })
    );
  }

  render() {
    const {
      navigator: navigatorName,
      scene: sceneName,
      children,
      overlay,
      style,
      ...props
    } = this.props;
    const { loading } = this.state;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    const id = toId(navigatorName, sceneName);
    const pass = { loading, id };

    return (
      <>
        {overlay && (
          <Animated.View
            style={{
              position: 'absolute',
              height: '100%',
              width: '100%',
              overflow: 'hidden',
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 0, 1],
                    outputRange: [height, 0, 0],
                  }),
                },
              ],
              backgroundColor: scene.active.value.interpolate({
                inputRange: [0, 1],
                outputRange: ['rgba(0,0,0,0)', 'rgba(0,0,0,0.4)'],
              }),
            }}
          />
        )}
        <Animated.View
          {...props}
          style={[
            {
              position: 'absolute',
              width: '100%',
              height: '100%',
              backgroundColor: 'white',
              overflow: 'hidden',
            },
            {
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 1],
                    outputRange: [height, 0],
                  }),
                },
              ],
            },
            style,
          ]}
        >
          {typeof children === 'function' ? children(pass) : children}
        </Animated.View>
      </>
    );
  }
}
