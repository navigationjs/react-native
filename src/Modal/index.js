import Navigator from './Navigator';
import Scene from './Scene';
import Wrap from './Wrap';

export default {
  Navigator,
  Scene,
  Wrap,
};
