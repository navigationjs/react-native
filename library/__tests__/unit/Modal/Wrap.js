import React from 'react';
import { View } from 'react-native';
import { render } from '@testing-library/react-native';
import Modal from '../../../src/Modal';
import navigation from '../../../src/Navigation';

describe('Modal.Wrap', () => {
  it('should match snapshot', async () => {
    const navigator = new Modal.Navigator('main');
    navigator.addScenes(new Modal.Scene('first'));
    navigation.addNavigators(navigator);
    const { baseElement } = render(
      <Modal.Wrap navigator={'main'} scene={'first'}>
        {props => <View {...props} />}
      </Modal.Wrap>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render overlay view if overlay prop passed', () => {
    const navigator = new Modal.Navigator('main');
    navigator.addScenes(new Modal.Scene('first'));
    navigation.addNavigators(navigator);
    const { baseElement } = render(
      <Modal.Wrap overlay navigator={'main'} scene={'first'}>
        {props => <View {...props} />}
      </Modal.Wrap>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render backgroundColor transparent if transparent prop passed', () => {
    const navigator = new Modal.Navigator('main');
    navigator.addScenes(new Modal.Scene('first'));
    navigation.addNavigators(navigator);
    const { baseElement } = render(
      <Modal.Wrap transparent navigator={'main'} scene={'first'}>
        {props => <View {...props} />}
      </Modal.Wrap>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should remove all listeners from scene active on unmount', () => {
    const navigator = new Modal.Navigator('navigator');
    const scene = new Modal.Scene('scene');
    navigator.addScenes(scene);
    navigation.addNavigators(navigator);

    const { unmount } = render(
      <Modal.Wrap navigator={'navigator'} scene={'scene'}>
        {props => <View {...props} />}
      </Modal.Wrap>
    );

    scene.active.value = {
      removeAllListeners: jest.fn(),
    };
    unmount();
    expect(scene.active.value.removeAllListeners).toBeCalled();
  });

  describe('loading prop', () => {
    it('should be true by default', async () => {
      const navigator = new Modal.Navigator('main');
      navigator.addScenes(new Modal.Scene('first'));
      navigation.addNavigators(navigator);

      const { getByTestId } = render(
        <Modal.Wrap navigator={'main'} scene={'first'}>
          {props => <View testID={'scene'} {...props} />}
        </Modal.Wrap>
      );

      const component = getByTestId('scene');

      expect(component.props.loading).toBeTruthy();
    });

    it('should be false when scene is showed and true when hidden', async () => {
      const navigator = new Modal.Navigator('main');
      navigator.addScenes(new Modal.Scene('first'));
      navigation.addNavigators(navigator);

      const { getByTestId } = render(
        <Modal.Wrap navigator={'main'} scene={'first'}>
          {props => <View testID={'scene'} {...props} />}
        </Modal.Wrap>
      );

      const component = getByTestId('scene');

      await navigation.go('main', 'first');
      expect(component.props.loading).toBeFalsy();
      await navigation.back();
      expect(component.props.loading).toBeTruthy();
    });

    it('should pass id and loading props to children if children if function', () => {
      const navigator = new Modal.Navigator('main');
      navigator.addScenes(new Modal.Scene('first'));
      navigation.addNavigators(navigator);

      const { getByTestId } = render(
        <>
          <Modal.Wrap navigator={'main'} scene={'first'}>
            {props => <View testID={'scene1'} {...props} />}
          </Modal.Wrap>
          <Modal.Wrap navigator={'main'} scene={'first'}>
            <View testID={'scene2'} />}
          </Modal.Wrap>
        </>
      );

      const scene1 = getByTestId('scene1');
      expect(scene1.props.id).toBe('main/first');
      expect(scene1.props.loading).toBeTruthy();

      const scene2 = getByTestId('scene2');
      expect(scene2.props.id).not.toBeDefined();
      expect(scene2.props.loading).not.toBeDefined();
    });
  });
});
