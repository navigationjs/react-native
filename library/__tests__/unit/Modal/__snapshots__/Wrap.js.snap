// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.Wrap should match snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "height": "100%",
          "overflow": "hidden",
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        id="main/first"
        loading={true}
      />
    </View>
  </View>
</View>
`;

exports[`Modal.Wrap should render backgroundColor transparent if transparent prop passed 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "height": "100%",
          "overflow": "hidden",
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        id="main/first"
        loading={true}
      />
    </View>
  </View>
</View>
`;

exports[`Modal.Wrap should render overlay view if overlay prop passed 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.4)",
          "height": "100%",
          "opacity": 0,
          "overflow": "hidden",
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
          "height": "100%",
          "overflow": "hidden",
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 1334,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        id="main/first"
        loading={true}
      />
    </View>
  </View>
</View>
`;
